frontend/app/routes/_public+/apply+/$id+/demographics-part2.tsx

```diff
@@ -19,81 +25,187 @@ export const handle = {
   const sexAtBirthTypes = await getLookupService().getAllSexAtBirthTypes();
   const mouthPainTypes = await getLookupService().getAllMouthPaintTypes();
   const lastTimeDentistVisitTypes = await getLookupService().getAllLastTimeDentistVisitTypes();
   const avoidedDentalCostTypes = await getLookupService().getAllAvoidedDentalCostTypes();
+  const genderTypes = await getLookupService().getAllGenderTypes();
   const applyFlow = getApplyFlow();
   const { id, state } = await applyFlow.loadState({ request, params });
-  return json({ sexAtBirthTypes, mouthPainTypes, lastTimeDentistVisitTypes, avoidedDentalCostTypes, id, state });
+
+  const otherGenderCode = genderTypes.find((method) => method.id === OTHER_GENDER_TYPE_ID);
+  if (!otherGenderCode) {
+    throw new Response('Expected gender type!', { status: 500 });
+  }
+
+  return json({ otherGenderCode, genderTypes, sexAtBirthTypes, mouthPainTypes, lastTimeDentistVisitTypes, avoidedDentalCostTypes, id, state: state.demographicsPart2 });
 }
 
-export async function action({ request }: ActionFunctionArgs) {
-  return json('');
+export async function action({ request, params }: ActionFunctionArgs) {
+  const applyFlow = getApplyFlow();
+  const { id } = await applyFlow.loadState({ request, params });
+
+  const formSchema = z.object({
+    genderTypeChoosen: z.string({ required_error: 'empty-radio' }),
+    otherGenderField: z.string().min(1, { message: 'empty-field' }).optional(),
+    mouthPainType: z.string({ required_error: 'empty-radio' }),
+    lastDentalVisitType: z.string({ required_error: 'empty-radio' }),
+    avoidedDentalCareDueToCost: z.string({ required_error: 'empty-radio' }),
+  });
+
+  const formData = Object.fromEntries(await request.formData());
+  const parsedDataResult = formSchema.safeParse(formData);
+
+  if (!parsedDataResult.success) {
+    return json({
+      errors: parsedDataResult.error.format(),
+      formData: formData as Partial<z.infer<typeof formSchema>>,
+    });
+  }
+
+  const sessionResponseInit = await applyFlow.saveState({
+    request,
+    params,
+    state: { demographicsPart2: parsedDataResult.data },
+  });
+
+  return redirect(`/apply/${id}/review-information`, sessionResponseInit);
 }
 
 export default function DemographicsPart2() {
-  const { sexAtBirthTypes, mouthPainTypes, lastTimeDentistVisitTypes, avoidedDentalCostTypes, id } = useLoaderData<typeof loader>();
+  const { otherGenderCode, genderTypes, mouthPainTypes, lastTimeDentistVisitTypes, avoidedDentalCostTypes, id, state } = useLoaderData<typeof loader>();
   const { i18n, t } = useTranslation(i18nNamespaces);
+  const [otherGenderChecked, setOtherGenderChecked] = useState(state?.genderTypeChoosen === otherGenderCode.id);
+  const actionData = useActionData<typeof action>();
+  const errorSummaryId = 'error-summary';
+
+  const otherGenderHandler = () => {
+    setOtherGenderChecked(true);
+  };
+
+  const nonOtherGenderHandler = () => {
+    setOtherGenderChecked(false);
+  };
```

Comment:
For consistency, consider not mixing arrow functions with normal functions unless absolutely necessary

```suggestion
  function otherGenderHandler() {
    setOtherGenderChecked(true);
  };

  function nonOtherGenderHandler() {
    setOtherGenderChecked(false);
  };
```

Reply:
arrows removed